README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
multifwf
========

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/multifwf)](http://cran.r-project.org/package=multifwf) [![Build Status](https://travis-ci.org/prontog/multifwf.svg)](https://travis-ci.org/prontog/multifwf) [![codecov](https://codecov.io/gh/prontog/multifwf/branch/master/graphs/badge.svg)](https://codecov.io/gh/prontog/multifwf)

Read a table of fixed width formatted data of different types into a data.frame for each type. This package is for people that need to load a file that contains lines of different fixed with format. Think of it as an extension to the **read.fwf** function. One use-case is reading a log file containing messages of a fixed-width text protocol.

An example
----------

``` r
library(multifwf)

# Create a temp file with a few lines from a simple exchange protocol.
ff <- tempfile()
cat(file = ff, 
    '10:15:03:279NOSLMT0000666    EVILCORP00010.77SomeClientId    SomeAccountId   ',
    '10:15:03:793OC000001BLMT0000666    EVILCORP00010.77SomeClientId    SomeAccountId   ',
    '10:17:45:153NOBLMT0000666    EVILCORP00001.10AnotherClientId AnotherAccountId',
    '10:17:45:487RJAnotherClientId 004price out of range                              ',
    '10:18:28:045NOBLMT0000666    EVILCORP00011.00AnotherClientId AnotherAccountId',
    '10:18:28:472OC000002BLMT0000666    EVILCORP00011.00AnotherClientId AnotherAccountId',
    '10:18:28:642TR0000010000010000666    EVILCORP00010.77',
    '10:18:28:687TR0000010000020000666    EVILCORP00010.77', 
    sep = '\n')

# Create a list of specs. Each item contains the specification for each message
# type of this simple protocol.
specs <- list()
specs[['newOrder']] = data.frame(widths    = c(12, 2, 1, 3, 7, 
                                                12, 8, 16, 16), 
                                  col.names = c('timestamp', 'msgType', 'side', 'type', 'volume', 
                                                'symbol', 'price', 'clientId', 'accountId'))
specs[['orderConf']] = data.frame(widths   = c(12, 2, 6, 1, 3,
                                                7, 12, 8, 16, 16), 
                                  col.names = c('timestamp', 'msgType', 'orderId', 'side', 'type', 
                                                'volume', 'symbol', 'price', 'clientId', 'accountId'))

specs[['rejection']] = data.frame(widths    = c(12, 2, 16, 3, 48), 
                                  col.names = c('timestamp', 'msgType', 'clientId', 'rejectionCode', 'text'))

specs[['trade']] = data.frame(widths   = c(12, 2, 6, 6, 7,
                                           12, 8), 
                              col.names = c('timestamp', 'msgType', 'tradeId', 'orderId', 'volume', 
                                            'symbol', 'price'))

# The selector function is responsible for identifying the message type of a line.
myselector <- function(line, specs) {
    s <- substr(line, 13, 14)
    spec_name = ''
    if (s == 'NO')
        spec_name = 'newOrder'
    else if (s == 'OC')
        spec_name = 'orderConf'
    else if (s == 'TR')
        spec_name = 'trade'
    else if (s == 'RJ')
        spec_name = 'rejection'

    spec_name
}

read.multi.fwf(ff, multi.specs = specs, select = myselector)
#> $newOrder
#>      timestamp msgType side type volume       symbol price
#> 1 10:15:03:279      NO    S  LMT    666     EVILCORP 10.77
#> 2 10:17:45:153      NO    B  LMT    666     EVILCORP  1.10
#> 3 10:18:28:045      NO    B  LMT    666     EVILCORP 11.00
#>           clientId        accountId
#> 1 SomeClientId     SomeAccountId   
#> 2 AnotherClientId  AnotherAccountId
#> 3 AnotherClientId  AnotherAccountId
#> 
#> $orderConf
#>      timestamp msgType orderId side type volume       symbol price
#> 1 10:15:03:793      OC       1    B  LMT    666     EVILCORP 10.77
#> 2 10:18:28:472      OC       2    B  LMT    666     EVILCORP 11.00
#>           clientId        accountId
#> 1 SomeClientId     SomeAccountId   
#> 2 AnotherClientId  AnotherAccountId
#> 
#> $rejection
#>      timestamp msgType         clientId rejectionCode
#> 1 10:17:45:487      RJ AnotherClientId              4
#>                                               text
#> 1 price out of range                              
#> 
#> $trade
#>      timestamp msgType tradeId orderId volume       symbol price
#> 1 10:18:28:642      TR       1       1    666     EVILCORP 10.77
#> 2 10:18:28:687      TR       1       2    666     EVILCORP 10.77

unlink(ff)
```

Installation
------------

The easiest way to install `multifwf` is from CRAN with `install.packages('devtools')`.

To install directly from this repo you can use the [devtools](https://github.com/hadley/devtools) package.

You can also clone the repo and build it yourself. The easiest way to do so is by opening *multifwf.Rproj* with *RStudio*. To build the package you will also need:

-   roxygen2 (for the documentation)
-   testthat (to test the package)
-   rstudio (optional)


NEWS.md
multifwf 0.5.0
============
* Added read_multi_fwf that uses readr::read_fwf to parse the files. This is usually considerably faster than read.fwf.

multifwf 0.2.2
============
* Added importFrom(utils,read.fwf) to NAMESPACE.
* Added 'Imports: utils' to DECSRIPTION file.
* Added dependency on R versions >= 3.1.1. Just to be safe since I've never tested it on an earlier version.

multifwf 0.2.1
============
* Added README.Rmd to generate the final README.md.
* Added a better example.
* Followed Hadley Wickham's advice on [releasing a package](http://r-pkgs.had.co.nz/). Thanks Hadley!

multifwf 0.2
============
* fixed: Warnings on connections that are left open.

multifwf 0.1
============
* First version
* known bug: Warnings on connections that are left open.


tests/testthat.R
library(testthat)
library(multifwf)
test_check('multifwf')


tests/testthat/test_read.multi.fwf.R
##### Helper Functions ##############################################
getSpecs <- function() {
    specs <- list()
    specs[['sp1']] <- data.frame(widths = c(1, 2, 3),
                                  col.names = c('Col1', 'Col2', 'Col3'))
    specs[['sp2']] <- data.frame(widths = c(3, 2, 1),
                                  col.names = c('C1', 'C2', 'C3'))
    specs
}

selector <- function(line, specs) {
    s <- substr(line, 1, 1)
    spec_name <- ''
    if (s == '1')
        spec_name <- 'sp1'
    else if (s == '2')
        spec_name <- 'sp2'

    spec_name
}

emptySelector <- function(line, specs) {
    ''
}
###################################################################

test_that("empty spec returns an empty list", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- list()

    data <- read.multi.fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    expect_equal(length(data), 0)
})

test_that("empty selector returns an empty list", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- getSpecs()

    data <- read.multi.fwf(ff, multi.specs = myspecs, select = emptySelector)
    unlink(ff)

    expect_equal(length(data), 2)
})

test_that("data and spec lists have the same length", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- getSpecs()

    data <- read.multi.fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    expect_equal(length(data), length(myspecs))
})

test_that("a file with no matches gives a data list with null elements", {
    ff <- tempfile()
    cat(file = ff, 'lalala', 'dododo', 'hehehe', 'hohoho', sep = '\n')

    myspecs <- getSpecs()

    data <- read.multi.fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    for (i in data) {
        expect_null(i)
    }
})

test_that("2 'sp1' and 1 'sp2'", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- getSpecs()

    data <- read.multi.fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    expect_equal(nrow(data$sp1), 2)
    expect_equal(nrow(data$sp2), 1)
})


tests/testthat/test_read_multi_fwf.R
##### Helper Functions ##############################################
getSpecs <- function() {
    specs <- list()
    specs[['sp1']] <- data.frame(widths = c(1, 2, 3),
                                  col_names = c('Col1', 'Col2', 'Col3'))
    specs[['sp2']] <- data.frame(widths = c(3, 2, 1),
                                  col_names = c('C1', 'C2', 'C3'))
    specs
}

selector <- function(line, specs) {
    s <- substr(line, 1, 1)
    spec_name <- ''
    if (s == '1')
        spec_name <- 'sp1'
    else if (s == '2')
        spec_name <- 'sp2'

    spec_name
}

emptySelector <- function(line, specs) {
    ''
}
###################################################################

test_that("empty spec returns an empty list", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- list()

    data <- read_multi_fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    expect_equal(length(data), 0)
})

test_that("empty selector returns an empty list", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- getSpecs()

    data <- read_multi_fwf(ff, multi.specs = myspecs, select = emptySelector)
    unlink(ff)

    expect_equal(length(data), 2)
})

test_that("data and spec lists have the same length", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- getSpecs()

    data <- read_multi_fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    expect_equal(length(data), length(myspecs))
})

test_that("a file with no matches gives a data list with null elements", {
    ff <- tempfile()
    cat(file = ff, 'lalala', 'dododo', 'hehehe', 'hohoho', sep = '\n')

    myspecs <- getSpecs()

    data <- read_multi_fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    for (i in data) {
        expect_null(i)
    }
})

test_that("2 'sp1' and 1 'sp2'", {
    ff <- tempfile()
    cat(file = ff, '123456', '287654', '987654', '198765', sep = '\n')

    myspecs <- getSpecs()

    data <- read_multi_fwf(ff, multi.specs = myspecs, select = selector)
    unlink(ff)

    expect_equal(nrow(data$sp1), 2)
    expect_equal(nrow(data$sp2), 1)
})


R/read_multi_fwf.R
#' Read Fixed Width Format Files containing lines of different Type
#' 
#' Read a table of fixed width formatted data of different types into a \link{tibble} for each type.
#' 
#' @param file either a path to a file, a connection, or literal data (either a single string or a raw vector). 
#' 
#' Files ending in .gz, .bz2, .xz, or .zip will be automatically uncompressed. Files starting with http://, https://, ftp://, or ftps:// will be automatically downloaded. Remote gz files can also be automatically downloaded and decompressed.
#' Literal data is most useful for examples and tests. It must contain at least one new line to be recognised as data (instead of a path) or be a vector of greater than length 1.
#' Using a value of clipboard() will read from the system clipboard.
#' @param multi.specs A named list of data.frames containing the following columns:
#'     \tabular{ll}{
#'         widths \tab   see \link{fwf_widths}\cr
#'         col_names \tab see \link{fwf_widths}\cr
#'     }
#' 
#' For more info on these fields see \code{\link{read_fwf}}.
#' 
#' Note that each list item should have a name. This is important for the select function.
#' @param select A function to select the type of a line. This selector should have parameters:
#'     \tabular{ll}{
#'         line \tab   the line\cr
#'         specs \tab  the multi.specs list that was passed to read.multi.fwf\cr
#'     }
#'     
#' The select function should return the name of the spec that matches the line. read.multi.fwf will then use this name to select the a spec from the passed multi.spec. This is why multi.spec should be a named list. 
#' If there is no match then NULL can be returned.
#' @param skip number of initial lines to skip; see \link{read_fwf}.
#' @param n the maximum number of records (lines) to be read, defaulting to no limit.
#' @param ... further arguments to be passed to \link{read_fwf}.
#' @return Return value is a named list with an item for each spec in multi.spec. If there was at least one line in file, matching a spec, then the named item will be a \link{tibble}. Otherwise it will be NULL.
#' @author
#' Panos Rontogiannis \email{panos@ronto.net}
#' @seealso \code{\link{read_fwf}}
#' @examples
#' ff <- tempfile()
#' cat(file = ff, '123456', '287654', '198765', sep = '\n')
#' specs <- list()
#' specs[['sp1']] = data.frame(widths = c(1, 2, 3), 
#'                             col_names = c('Col1', 'Col2', 'Col3'))
#' specs[['sp2']] = data.frame(widths = c(3, 2, 1), 
#'                             col_names = c('C1', 'C2', 'C3'))
#' 
#' myselector <- function(line, specs) {
#'     s <- substr(line, 1, 1)
#'     spec_name = ''
#'     if (s == '1')
#'         spec_name = 'sp1'
#'     else if (s == '2')
#'         spec_name = 'sp2'
#' 
#'     spec_name
#' }
#' 
#' read_multi_fwf(ff, multi.specs = specs, select = myselector)    
#' #> sp1: 1 23 456 \ 1 98 765, sp2: 287 65 4
#' 
#' unlink(ff)
#' @export
read_multi_fwf <- function(file,
                           multi.specs,
                           select,
                           skip = 0,
                           n = -1,
                           ...) {
    ##### Helper functions #############################
    
    # Convert each spec into a list containing a SPEC field. Also add
    # the FILENAME field for the temp file where all lines of this type (spec)
    # will be stored (temporarily) for read.fwf.
    prepareAsList <- function(s) {
        s <- list(SPEC = s)
        s$FILENAME <- tempfile("Rmfwf.")
        return(s)
    }
    
    # Parse a line and write to the temp file of the matching spec.
    doone <- function(line) {
        spec_name <- select(line, multi.specs)
        s <- extended.specs[[spec_name]]
        if (is.null(s))
            return()
        cat(file = s$FILENAME, line, "\n", append = TRUE)
        invisible()
    }
    
    # Read Fixed-Width Format temp file.
    readFwf <- function(s) {
        fi <- file.info(s$FILENAME)
        if (!is.na(fi$size) & fi$size > 0) {
            s$Data <- readr::read_fwf(file = s$FILENAME,
                               col_positions = readr::fwf_widths(widths = s$SPEC$widths,
                                                                 col_names = s$SPEC$col_names),
                               ...)
        }
        else {
            #s$Data <- NA
            s <- NULL
        }
        return(s)
    }
    
    # Prepare an element for return.
    prepareRetval <- function(s) {
        s <- s$Data
        return(s)
    }
    ####################################################
    
    extended.specs <- lapply(multi.specs, FUN = prepareAsList)
    
    if (is.character(file)) {
        file <- file(file, "rt")
        on.exit(close(file), add = TRUE)
    }
    else if (!isOpen(file)) {
        open(file, "rt")
        on.exit(close(file), add = TRUE)
    }
    
    # Skip lines.
    if (skip)
        readLines(file, n = skip)
    
    repeat ({
        if (n == 0L)
            break
        if (n == -1L)
            n <- 16
        
        raw <- readLines(file, n = n)
        nread <- length(raw)
        if (nread == 0)
            break
        
        lapply(raw, FUN = doone)
    })
    
    extended.specs <- lapply(extended.specs, FUN = readFwf)
    loaded.data <- lapply(extended.specs, FUN = prepareRetval)
    
    return(loaded.data)
}


R/read.multi.fwf.R
#' Read Fixed Width Format Files containing lines of different Type
#' 
#' Read a table of fixed width formatted data of different types into a \link{data.frame} for each type.
#' 
#' @param file the name of the file which the data are to be read from.
#' 
#' Alternatively, file can be a \link{connection}, which will be opened if necessary, and if so closed at the end of the function call.
#' @param multi.specs A named list of data.frames containing the following columns:
#'     \tabular{ll}{
#'         widths \tab   see \link{read.fwf}\cr
#'         col.names \tab see \link{read.table}\cr
#'         row.names \tab see \link{read.table}\cr
#'     }
#' 
#' For more info on these fields see \code{\link{read.fwf}}.
#' 
#' Note that each list item should have a name. This is important for the select function.
#' @param select A function to select the type of a line. This selector should have parameters:
#'     \tabular{ll}{
#'         line \tab   the line\cr
#'         specs \tab  the multi.specs list that was passed to read.multi.fwf\cr
#'     }
#'     
#' The select function should return the name of the spec that matches the line. read.multi.fwf will then use this name to select the a spec from the passed multi.spec. This is why multi.spec should be a named list. 
#' If there is no match then NULL can be returned.
#' @param header a logical value indicating whether the file contains the names of the variables as its first line. If present, the names must be delimited by sep.
#' @param sep character; the separator used internally; should be a character that does not occur in the file (except in the header).
#' @param skip number of initial lines to skip; see \link{read.fwf}.
#' @param n the maximum number of records (lines) to be read, defaulting to no limit.
#' @param buffersize Maximum number of lines to read at one time
#' @param ... further arguments to be passed to \link{read.fwf}.
#' @return Return value is a named list with an item for each spec in multi.spec. If there was at least one line in file, matching a spec, then the named item will be a \link{data.frame}. Otherwise it will be NULL.
#' @author
#' Panos Rontogiannis \email{p.g.ronto@@gmail.com}
#' @seealso \code{\link{read.fwf}}
#' @examples
#' ff <- tempfile()
#' cat(file = ff, '123456', '287654', '198765', sep = '\n')
#' specs <- list()
#' specs[['sp1']] = data.frame(widths = c(1, 2, 3), 
#'                             col.names = c('Col1', 'Col2', 'Col3'))
#' specs[['sp2']] = data.frame(widths = c(3, 2, 1), 
#'                             col.names = c('C1', 'C2', 'C3'))
#' 
#' myselector <- function(line, specs) {
#'     s <- substr(line, 1, 1)
#'     spec_name = ''
#'     if (s == '1')
#'         spec_name = 'sp1'
#'     else if (s == '2')
#'         spec_name = 'sp2'
#' 
#'     spec_name
#' }
#' 
#' read.multi.fwf(ff, multi.specs = specs, select = myselector)    
#' #> sp1: 1 23 456 \ 1 98 765, sp2: 287 65 4
#' 
#' unlink(ff)
#' @export
read.multi.fwf <- function(file,
                           multi.specs,
                           select,
                           header = FALSE,
                           sep = "\t",
                           skip = 0,
                           n = -1,
                           buffersize = 2000,
                           ...) {
    ##### Helper functions #############################

    # Convert each spec into a list containing a SPEC field. Also add
    # the FILENAME field for the temp file where all lines of this type (spec)
    # will be stored (temporarily) for read.fwf.
    prepareAsList <- function(s) {
        s <- list(SPEC = s)
        s$FILENAME <- tempfile("Rmfwf.")
        return(s)
    }

    # Add header line to a temp file.
    addHeader <- function(s, headerline) {
        cat(file = s$FILENAME, headerline, "\n")
        return(s)
    }

    # Parse a line and write to the temp file of the matching spec.
    doone <- function(line) {
        spec_name <- select(line, multi.specs)
        s <- extended.specs[[spec_name]]
        if (is.null(s))
            return()
        cat(file = s$FILENAME, line, "\n", append = TRUE)
        invisible()
    }

    # Read Fixed-Width Format temp file.
    readFwf <- function(s) {
        fi <- file.info(s$FILENAME)
        if (!is.na(fi$size) & fi$size > 0) {
            s$Data <- read.fwf(file = s$FILENAME,
                               widths = s$SPEC$widths,
                               header = header,
                               sep = sep,
                               row.names = s$SPEC$row.names,
                               col.names = s$SPEC$col.names,
                               ...)
        }
        else {
            #s$Data <- NA
            s <- NULL
        }
        return(s)
    }

    # Prepare an element for return.
    prepareRetval <- function(s) {
        s <- s$Data
        return(s)
    }
    ####################################################

    extended.specs <- lapply(multi.specs, FUN = prepareAsList)

    if (is.character(file)) {
        file <- file(file, "rt")
        on.exit(close(file), add = TRUE)
    }
    else if (!isOpen(file)) {
        open(file, "rt")
        on.exit(close(file), add = TRUE)
    }

    # Handle header line.
    if (header) {
        headerline <- readLines(file, n = 1L)
        lapply(extended.specs, FUN = addHeader, headerline)
    }
    
    # Skip lines.
    if (skip)
        readLines(file, n = skip)

    repeat ({
        if (n == 0L)
            break
        if (n == -1L)
            n <- 16

        raw <- readLines(file, n = n)
        nread <- length(raw)
        if (nread == 0)
            break

        lapply(raw, FUN = doone)
    })

    extended.specs <- lapply(extended.specs, FUN = readFwf)
    loaded.data <- lapply(extended.specs, FUN = prepareRetval)

    return(loaded.data)
}
